### PR TITLE
Request for Integrating the new NAS algorithm: Cream

### PR DESCRIPTION
**Pull request by penghouwen**
Cream of the Crop: Distilling Prioritized Paths For One-Shot Neural Architecture Search

### CODE REVIEW

File: examples/nas/cream/Cream.md [hunk 1]
```diff
@@ -0,0 +1,94 @@
+# Cream of the Crop: Distilling Prioritized Paths For One-Shot Neural Architecture Search
```
Comment (ultmaster) [inline examples/nas/cream/Cream.md:1]: Please put a link to documentation here instead of copying the full file. Refer to other examples.

File: examples/nas/cream/requirements.txt [hunk 2]
```diff
@@ -0,0 +1,12 @@
+yacs
+numpy==1.17
+opencv-python==4.0.1.24
+torchvision==0.2.1
+thop
+git+https://github.com/sovrasov/flops-counter.pytorch.git
```
Comment (ultmaster) [inline examples/nas/cream/requirements.txt:6]: Can it be installed via ptflops?

File: examples/nas/cream/requirements.txt [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+pillow==6.1.0
+torch==1.2
+timm==0.1.20
+tensorboardx==1.2
+git+https://github.com/Tramac/torchscope.git
```
Comment (ultmaster) [inline examples/nas/cream/requirements.txt:11]: Can it be installed via pip torchscope?

File: examples/nas/cream/requirements.txt [hunk 4]
```diff
@@ -0,0 +1,12 @@
+yacs
+numpy==1.17
+opencv-python==4.0.1.24
+torchvision==0.2.1
+thop
+git+https://github.com/sovrasov/flops-counter.pytorch.git
+pillow==6.1.0
+torch==1.2
```
Comment (ultmaster) [inline examples/nas/cream/requirements.txt:8]: Can we upgrade to latest version and pytorch and remove tensorboardx dependency?

File: examples/nas/cream/requirements.txt [hunk 5]
```diff
@@ -0,0 +1,12 @@
+yacs
+numpy==1.17
+opencv-python==4.0.1.24
+torchvision==0.2.1
```
Comment (ultmaster) [inline examples/nas/cream/requirements.txt:4]: Can it be upgraded?

File: docs/en_US/NAS/Cream.md [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+* torch >= 1.2
+* torchscope
+* apex (not necessary, please make sure your nvcc CUDA version is the same with pytorch CUDA verision)
+
+## Data Preparation 
+You need to first download the [ImageNet-2012](http://www.image-net.org/) to the folder `./data/imagenet` and move the validation set to the subfolder `./data/imagenet/val`. To move the validation set, you cloud use the following script: <https://raw.githubusercontent.com/soumith/imagenetloader.torch/master/valprep.sh> 
+
+Put the imagenet data in ${Root}/data. It should be like following:
+```buildoutcfg
+${Root}/data/imagenet/train
```
Comment (ultmaster) [inline docs/en_US/NAS/Cream.md:29]: What is $Root here?

File: docs/en_US/NAS/Cream.md [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+## Quick Start
+
+### I. Search
+
+First, build environments for searching.
+```
+pip install -r ./examples/nas/cream/requirements.txt
+```
+
+To search for an architecture, you need to configure the parameters `flops_minimum` and `flops_maximum` to specify the desired model flops, such as [0,600]MB flops. You can specify the flops interval by changing these two parameters in `./examples/nas/cream/supernet.sh`
```
Comment (ultmaster) [inline docs/en_US/NAS/Cream.md:44]: supernet.sh is missing.

Comment (ultmaster) [inline docs/en_US/NAS/Cream.md:44]: Better to change working directory to examples. Then `./examples/nas/cream` is no longer necessary.

File: docs/en_US/NAS/Cream.md [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+```buildoutcfg
+--flops_minimum 0 # Minimum Flops of Architecture
+--flops_maximum 600 # Maximum Flops of Architecture
+```
+
+After you specify the flops of the architectures you would like to search, you can search an architecture now by running:
```
Comment (ultmaster) [inline docs/en_US/NAS/Cream.md:50]: Again missing

File: docs/en_US/NAS/Cream.md [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+```buildoutcfg
+sh ./experiments/scripts/supernet.sh
+
+```
+
+### II. Test
+To test our trained of models, you need to use `model_selection` in `./examples/nas/cream/test.sh` to specify which model to test.
```
Comment (ultmaster) [inline docs/en_US/NAS/Cream.md:57]: Working directory. same here.

File: examples/nas/cream/distributed_test.sh [hunk 10]
```diff
@@ -0,0 +1,4 @@
+#!/bin/bash
+NUM_PROC=$1
+shift
+python -m torch.distributed.launch --nproc_per_node=$NUM_PROC ./examples/nas/cream/test.py "$@"
```
Comment (ultmaster) [inline examples/nas/cream/distributed_test.sh:4]: Same here.

File: docs/en_US/NAS/Cream.md [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+
+## Examples
+
+[Example code](https://github.com/microsoft/nni/tree/master/examples/nas/cream)
+
+## Requirements
+* python >= 3.6
+* torch >= 1.2
+* torchscope
+* apex (not necessary, please make sure your nvcc CUDA version is the same with pytorch CUDA verision)
```
Comment (ultmaster) [inline docs/en_US/NAS/Cream.md:31]: Tested. Apex is actually mandatory.

File: examples/nas/cream/Cream.md (hunk 1 above)
Comment (penghouwen) [reply examples/nas/cream/Cream.md:1]: Updated. I will add a Arxiv link in the near future

File: examples/nas/cream/requirements.txt (hunk 5 above)
Comment (penghouwen) [reply examples/nas/cream/requirements.txt:4]: Yes, but I am not sure the results can be exactly reproduced. 
To ensure the results can be easily reproduced, we suggested to use the default versions, which are well tested.

File: examples/nas/cream/requirements.txt (hunk 2 above)
Comment (penghouwen) [reply examples/nas/cream/requirements.txt:6]: Updated using pip install ptflops

File: examples/nas/cream/requirements.txt (hunk 3 above)
Comment (mapleam) [reply examples/nas/cream/requirements.txt]: No, 'pip install torchscope' would throw an error.

File: examples/nas/cream/requirements.txt (hunk 4 above)
Comment (mapleam) [reply examples/nas/cream/requirements.txt:8]: tensorboardX can be removed.

Comment (mapleam) [reply examples/nas/cream/requirements.txt:8]: Can be upgraded to the latest version, but the results can not be exactly reproduced.

File: docs/en_US/NAS/Cream.md (hunk 6 above)
Comment (mapleam) [reply docs/en_US/NAS/Cream.md:39]: ${Root} indicates codes root, 
`xxxxxx/nni/examples/nas/cream/`

File: docs/en_US/NAS/Cream.md (hunk 7 above)
Comment (mapleam) [reply docs/en_US/NAS/Cream.md:54]: Fixed, 'supernet.sh' should be 'run.sh'

Comment (mapleam) [reply docs/en_US/NAS/Cream.md:54]: Fixed.

File: examples/nas/cream/requirements.txt (hunk 2 above)
Comment (penghouwen) [reply examples/nas/cream/requirements.txt:6]: The change will bring an error

File: docs/en_US/NAS/Cream.md [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+process, aiming to boost the convergence of individual models. We introduce the concept of prioritized path, which refers to the architecture candidates exhibiting superior performance during training. Distilling knowledge from the prioritized
+paths is able to boost the training of subnetworks. Since the prioritized paths are changed on the fly depending on their performance and complexity, the final obtained paths are the cream of the crop. We directly select the most promising
+one from the prioritized paths as the final architecture, without using other complex search methods, such as reinforcement learning or evolution algorithms. The experiments on ImageNet verify such path distillation method can improve the
+convergence ratio and performance of the hypernetwork, as well as boosting the training of subnetworks. The discovered architectures achieve superior performance compared to the recent MobileNetV3 and EfficientNet families under aligned
+settings. Moreover, the experiments on object detection and more challenging search space show the generality and robustness of the proposed method.
+For more details, please refer to the paper (coming soon).
```
Comment (scarlett2018) [inline docs/en_US/NAS/Cream.md:11]: Do we want to release this after the paper is published?

File: docs/en_US/NAS/Cream.md [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+process, aiming to boost the convergence of individual models. We introduce the concept of prioritized path, which refers to the architecture candidates exhibiting superior performance during training. Distilling knowledge from the prioritized
+paths is able to boost the training of subnetworks. Since the prioritized paths are changed on the fly depending on their performance and complexity, the final obtained paths are the cream of the crop. We directly select the most promising
+one from the prioritized paths as the final architecture, without using other complex search methods, such as reinforcement learning or evolution algorithms. The experiments on ImageNet verify such path distillation method can improve the
+convergence ratio and performance of the hypernetwork, as well as boosting the training of subnetworks. The discovered architectures achieve superior performance compared to the recent MobileNetV3 and EfficientNet families under aligned
```
Comment (scarlett2018) [inline docs/en_US/NAS/Cream.md:9]: we show have the compare results present here to attract usage.

**Review by scarlett2018 — DISMISSED** [review]

File: docs/en_US/NAS/Cream.md [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+
+First, build environments for searching.
+
+```
+pip install -r ./requirements.txt
+```
+
+To search for an architecture, you need to configure the parameters `flops_minimum` and `flops_maximum` to specify the desired model flops, such as [0,600]MB flops. You can specify the flops interval by changing these two parameters in `./run.sh`
+
+```
```
Comment (scarlett2018) [inline docs/en_US/NAS/Cream.md:61]: like other NNI examples, do we want to share the search space and example code? https://nni.readthedocs.io/en/latest/NAS/ENAS.html

File: docs/en_US/NAS/Cream.md (hunk 12 above)
Comment (penghouwen) [reply docs/en_US/NAS/Cream.md:11]: Sure, I will update it soon.

File: docs/en_US/NAS/Cream.md (hunk 13 above)
Comment (penghouwen) [reply docs/en_US/NAS/Cream.md]: Updated.

File: docs/en_US/NAS/Cream.md (hunk 14 above)
Comment (penghouwen) [reply docs/en_US/NAS/Cream.md:67]: @scarlett2018  updated. pls have a check.  #2705

**Comment by penghouwen** [comment]
Hi @scarlett2018 , the issues had been solved, pls have a review.

File: docs/en_US/NAS/Cream.md [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+| Model (M Flops) | NNI (8Gpus) | Paper (16Gpus) | 
+| ---- |:-------------:| :-----:|
+| 14M | testing | 59.6 |
+| 42M | 65.8 | 66.5 |
+| 114M | 72.1 | 72.8 |
+| 285M | 76.7 | 77.6 |
+| 470M | 78.9 | 79.2 |
+| 600M | 79.4 | 80.0 |
+
+Top-1 Accuracy on ImageNet reported in the paper, supassing MobileNetV3 and EfficientNet-B0/B1.
```
Comment (QuanluZhang) [inline docs/en_US/NAS/Cream.md:23]: supassing -> surpassing

File: docs/en_US/NAS/Cream.md [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+```
+
+To search for an architecture, you need to configure the parameters `flops_minimum` and `flops_maximum` to specify the desired model flops, such as [0,600]MB flops. You can specify the flops interval by changing these two parameters in `./run.sh`
+
+```
+--flops_minimum 0 # Minimum Flops of Architecture
+--flops_maximum 600 # Maximum Flops of Architecture
+```
+
+For example, if you expect to search an architecture with model Flops <= 200M, please set the `flops_minimum` and `flops_maximum` to be `0` and `200`.
```
Comment (QuanluZhang) [inline docs/en_US/NAS/Cream.md:72]: better to use the same style for `Flops`, `flops`

File: docs/en_US/NAS/Cream.md [hunk 17]
```diff
@@ -0,0 +1,117 @@
...
+
+After you specify the flops of the architectures you would like to search, you can search an architecture now by running:
+
+```
+sh ./run.sh
+```
+
+Searched model needs to be retrained to obtain the final model. Retraining code will be released soon.
```
Comment (QuanluZhang) [inline docs/en_US/NAS/Cream.md:80]: better to explain how to get the final model, it is a json file or other format?

File: examples/nas/cream/distributed_test.sh (hunk 10 above)
Comment (QuanluZhang) [inline examples/nas/cream/distributed_test.sh:4]: suggest to use python3 instead of python

File: src/sdk/pynni/nni/nas/pytorch/cream/mutator.py [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+    flops_lb : number
+        Lower bound of flops.
+    flops_ub : number
+        Upper bound of flops.
+    flops_bin_num : number
+        Number of bins divided for the interval of flops to ensure the uniformity. Bigger number will be more
+        uniform, but the sampling will be slower.
+    flops_sample_timeout : int
+        Maximum number of attempts to sample before giving up and use a random candidate.
+    """
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/nas/pytorch/cream/mutator.py:31]: the `Parameters` does not match the arguments of this class

File: src/sdk/pynni/nni/nas/pytorch/cream/mutator.py [hunk 19]
```diff
@@ -0,0 +1,66 @@
...
+
+    def __init__(self, model, how_to_prob='even', pre_prob=(0.05, 0.05, 0.2, 0.4, 0.2, 0.1), CHOICE_NUM=6,
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/nas/pytorch/cream/mutator.py:33]: better to make `CHOICE_NUM` configurable

Comment (QuanluZhang) [reply src/sdk/pynni/nni/nas/pytorch/cream/mutator.py:33]: also better to expose `pre_prob` to be configured by users

File: src/sdk/pynni/nni/nas/pytorch/cream/mutator.py [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+            return None
+        elif self.how_to_prob == 'pre_prob':
+            return self.pre_prob
+        else:
+            raise ValueError("prob method not supported")
+
+    def sample_search(self):
+        """
+        Sample a candidate for training. When ``flops_func`` is not None, candidates will be sampled uniformly
+        relative to flops.
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/nas/pytorch/cream/mutator.py:53]: a little confused, where is `flops_func`?

File: src/sdk/pynni/nni/nas/pytorch/cream/trainer.py [hunk 21]
```diff
@@ -0,0 +1,308 @@
...
+    workers: int
+        Number of threads for data preprocessing. Not used for this trainer. Maybe removed in future.
+    device : torch.device
+        Device object. Either ``torch.device("cuda")`` or ``torch.device("cpu")``. When ``None``, trainer will
+        automatic detects GPU and selects GPU first.
+    log_frequency : int
+        Number of mini-batches to log metrics.
+    callbacks : list of Callback
+        Callbacks to plug into the trainer. See Callbacks.
+    """
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/nas/pytorch/cream/trainer.py:50]: does not match

File: src/sdk/pynni/nni/nas/pytorch/cream/trainer.py [hunk 22]
```diff
@@ -0,0 +1,308 @@
...
+                                                   'cuda', log_frequency, None)
+        self.train_loader = train_loader
+        self.valid_loader = valid_loader
+        self.log_frequency = log_frequency
+        self.batch_size = batch_size
+        self.mutator = mutator
+        self.optimizer = optimizer
+        self.model = model
+        self.loss = loss
+        self.est = est
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/nas/pytorch/cream/trainer.py:69]: what does `est` do?

File: src/sdk/pynni/nni/nas/pytorch/cream/trainer.py [hunk 23]
```diff
@@ -0,0 +1,308 @@
...
+            loss.backward()
+            self.optimizer.step()
+
+            prec1, prec5 = self.accuracy(output, target, topk=(1, 5))
+            metrics = {"prec1": prec1, "prec5": prec5, "loss": loss}
+            metrics = self.reduce_metrics(metrics, self.distributed)
+            meters.update(metrics)
+
+            if epoch > self.meta_sta_epoch and (
+                    (len(self.best_children_pool) < self.pool_size) or (prec1 > self.best_children_pool[-1][1] + 5) or
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/nas/pytorch/cream/trainer.py:265]: what is `self.best_children_pool[-1][1]`?

File: src/sdk/pynni/nni/nas/pytorch/cream/trainer.py [hunk 24]
```diff
@@ -0,0 +1,308 @@
...
+    log_frequency : int
+        Number of mini-batches to log metrics.
+    callbacks : list of Callback
+        Callbacks to plug into the trainer. See Callbacks.
+    """
+
+    def __init__(self, model, loss,
+                 optimizer, num_epochs, train_loader, valid_loader,
+                 mutator=None, batch_size=64, log_frequency=None,
+                 est=None, meta_sta_epoch=20, update_iter=200, slices=2, pool_size=10,
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/nas/pytorch/cream/trainer.py:55]: `est`: estimation, used for estimate model flops through a lookup table; type: obj (has two dicts). how to produce `est`, current layer with current resolution;

File: src/sdk/pynni/nni/nas/pytorch/cream/mutator.py (hunk 18 above)
Comment (QuanluZhang) [reply src/sdk/pynni/nni/nas/pytorch/cream/mutator.py:31]: the docstring of this class is not correct

File: docs/en_US/NAS/Cream.md [hunk 25]
```diff
@@ -0,0 +1,120 @@
...
+INPUT_ARCH: [[0], [3], [3, 3], [3, 1, 3], [3, 3, 3, 3], [3, 3, 3], [0]] # Random Architectures
+......
+```
+
+After adding `MODEL_SELECTION` in `./configs/retrain.yaml`, you need to use the following command to train the model.
+```buildoutcfg
+python -m torch.distributed.launch --nproc_per_node=8 ./retrain.py --cfg ./configs/retrain.yaml
+```
+
+### II. Test
```
Comment (QuanluZhang) [inline docs/en_US/NAS/Cream.md:96]: II -> III

**Review by ultmaster — APPROVED** [review]

File: docs/en_US/NAS/Cream.md (hunk 25 above)
Comment (penghouwen) [reply docs/en_US/NAS/Cream.md]: Solved.

**Review by QuanluZhang — APPROVED** [review]
